tests/test_utils.py

```diff
@@ -346,3 +347,26 @@ def test_get_compile_command_sort_args(tmpdir_cwd):
             "--no-annotate --no-emit-index-url --no-emit-trusted-host "
             "requirements.in setup.py"
         )
+
+
+@pytest.mark.parametrize(
+    ("values", "key"),
+    (
+        pytest.param(
+            ("foo", "bar", "baz", "qux", "quux"), lambda s: s[0], id="with key function"
+        ),
+        pytest.param(
+            (("f", "foo"), ("b", "bar"), ("b", "baz"), ("q", "qux"), ("q", "quux")),
+            None,
+            id="without key function",
+        ),
+    ),
+)
+def test_lookup_table(values, key):
+    expected = {"b": {"bar", "baz"}, "f": {"foo"}, "q": {"quux", "qux"}}
+    assert lookup_table(values, key) == expected
+
+
+def test_lookup_table_requires_key():
+    with pytest.raises(AssertionError, match="^key function must be specified$"):
```

Comment:
:exploding_head: Please don't use `AssertionError` outside of the test files as an API, it's hugely unreliable and you're not testing what you think you do.
Also, it's a good tone/convention to use r-strings for regexes.

Reply:
Okay, thanks. Addressed in 1c125852462698c1d40e50f796f81c73f3411f2b.